prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -152,81 +154,88 @@ const (
+		// TODO(fejta): do not require the json file https://github.com/kubernetes/test-infra/issues/16489
+		// As far as I can tell, there is no sane way to get these values other than just
+		// reading them out of the JSON file ourselves.
+		f, err := os.Open(af.gcsCredsFile)
+		if err != nil {
+			return "", err
+		}
+		defer f.Close()
+		auth := struct {
+			Type        string `json:"type"`
+			PrivateKey  string `json:"private_key"`
+			ClientEmail string `json:"client_email"`
+		}{}
+		if err := json.NewDecoder(f).Decode(&auth); err != nil {
+			return "", err
+		}
+		if auth.Type != "service_account" {
+			return "", fmt.Errorf("only service_account GCS auth is supported, got %q", auth.Type)
+		}
+		return storage.SignedURL(bucket, obj, &storage.SignedURLOptions{
+			Method:         "GET",
+			Expires:        time.Now().Add(10 * time.Minute),
+			GoogleAccessID: auth.ClientEmail,
+			PrivateKey:     []byte(auth.PrivateKey),
+		})
 	}
-	return storage.SignedURL(bucket, obj, &storage.SignedURLOptions{
-		Method:         "GET",
-		Expires:        time.Now().Add(10 * time.Minute),
-		GoogleAccessID: auth.ClientEmail,
-		PrivateKey:     []byte(auth.PrivateKey),
-	})
+	return af.opener.SignedURL(ctx, fmt.Sprintf("%s%s/%s", linkPrefix, bucket, obj))
 }
 
 type gcsArtifactHandle struct {
-	*storage.ObjectHandle
+	pkgio.Opener
+	Name string
 }
 
 func (h *gcsArtifactHandle) NewReader(ctx context.Context) (io.ReadCloser, error) {
-	return h.ObjectHandle.NewReader(ctx)
+	return h.Opener.Reader(ctx, h.Name)
 }
 
 func (h *gcsArtifactHandle) NewRangeReader(ctx context.Context, offset, length int64) (io.ReadCloser, error) {
-	return h.ObjectHandle.NewRangeReader(ctx, offset, length)
+	return h.Opener.RangeReader(ctx, h.Name, offset, length)
+}
+
+func (h *gcsArtifactHandle) Attrs(ctx context.Context) (pkgio.Attributes, error) {
+	return h.Opener.Attributes(ctx, h.Name)
 }
 
 // Artifact constructs a GCS artifact from the given GCS bucket and key. Uses the golang GCS library
 // to get read handles. If the artifactName is not a valid key in the bucket a handle will still be
 // constructed and returned, but all read operations will fail (dictated by behavior of golang GCS lib).
-func (af *GCSArtifactFetcher) Artifact(key string, artifactName string, sizeLimit int64) (api.Artifact, error) {
+func (af *GCSArtifactFetcher) Artifact(ctx context.Context, key string, artifactName string, sizeLimit int64) (api.Artifact, error) {
 	src, err := newGCSJobSource(key)
 	if err != nil {
 		return nil, fmt.Errorf("Failed to get GCS job source from %s: %v", key, err)
 	}
 
 	bucketName, prefix := extractBucketPrefixPair(src.jobPath())
-	bkt := af.client.Bucket(bucketName)
 	objName := path.Join(prefix, artifactName)
-	obj := &gcsArtifactHandle{bkt.Object(objName)}
-	signedURL, err := af.signURL(bucketName, objName)
+	obj := &gcsArtifactHandle{Opener: af.opener, Name: fmt.Sprintf("%s%s/%s", src.linkPrefix, src.bucket, objName)}
```

Comment:
In a follow-up, we should probably rename the `gcsArtifactHandle` to something neutral as well to make reading the code less confusing

Reply:
Also added above: https://github.com/kubernetes/test-infra/pull/17183#issue-402002610
Just didn't rename it already because the diff would be huge :)